sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -2452,9 +2491,16 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
           } else {
             values(i).getText
           }
+
           UTF8String.fromString(" " + v + " " + u)
         }
-        IntervalUtils.stringToInterval(UTF8String.concat(kvs: _*))
+        val intervalLiteral = UTF8String.concat(kvs: _*)
+        val calendarInterval = IntervalUtils.stringToInterval(intervalLiteral)
+        if (!conf.legacyIntervalEnabled && isYearMonthInterval && isDayTimeInterval) {
+          throw QueryParsingErrors.mixedIntervalUnitsError(intervalLiteral.toString.trim, ctx)
```

Comment:
it looks simpler to detect the mixed fields in https://github.com/apache/spark/pull/32949/files#diff-26fe6b511f4b8aedf5bdffae0d7c03aba69463c09fa4f881d77520125796e0faR2422

do we have a strong reason to do it here?

Reply:
Yes. it's to give users error message with literal which is easy to understand, as you [suggested](https://github.com/apache/spark/pull/32949#discussion_r664724125).
The assertion added is not for users but to detect a kind of bug.
`visitInterval` implicitly relies on the premise that `visitMultiUnitsInterval` is called and detect mixed fields.
So, if someone changes code around `visitInterval` in the future, the premise may not hold.
To detect such a case, I placed that assertion.

Reply:
I mean, can we do `throw QueryParsingErrors.mixedIntervalUnitsError` inside `visitInterval`? The check here adds some duplicated code and looks a bit convoluted.

Reply:
BTW we also check other errors in `parseIntervalLiteral`, which is the caller side of `visitMultiUnitsInterval`/`visitUnitToUnitInterval`. I think it's OK to check errors in the caller side if it makes the code simpler.

Reply:
At first I tried to do so but building literal string will happen twice (one in `visitMultiUnitsInterval` and one in `visitInterval`).
How about just removing the assertion here?
Or, is it acceptable to change `parseIntervalLiteral` and `visitMultiUnitsInterval` for the purpose?

Reply:
I'm confused. Where do we build the literal string in `parseIntervalLiteral`?